### PR TITLE
feat: add action to bump version

### PR DESCRIPTION
**Pull request by ddhorstman**
This is an action that auto-bumps the version of a package based on commit messages.

I started from https://github.com/phips28/gh-action-bump-version and added logic to support our desired use-case of applying version bumps when updating dependencies, as in https://github.com/ClipboardHealth/test-chassis. (test run [here](https://github.com/ClipboardHealth/test-chassis/actions/runs/3332992476/jobs/5514492491))

I authored the code in `index.js` from lines `114` to `163`. The rest is the author's original code, which I haven't modified.

### CODE REVIEW

**Comment by prudhvicbh** [comment]
The original repo[ had tests,](https://github.com/phips28/gh-action-bump-version/tree/master/tests/end-to-end) do we copy them along and also write tests for the changes made by us (lines 114 to 163) or is that a big ask?

File: .github/actions/auto-version-bump/index.js [hunk 1]
```diff
@@ -0,0 +1,333 @@
+// test
```
Comment (prudhvicbh) [inline .github/actions/auto-version-bump/index.js:1]: How do we test these action changes? Is this tested? I know it works but just asking

File: .github/actions/auto-version-bump/index.js [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+const { execSync, spawn } = require('child_process');
```
Comment (prudhvicbh) [inline .github/actions/auto-version-bump/index.js:2]: The original repo[ had tests,](https://github.com/phips28/gh-action-bump-version/tree/master/tests/end-to-end) do we copy them along and also write tests for the changes made by us (lines 114 to 163) or is that a big ask?

File: .github/actions/auto-version-bump/index.js [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+const { existsSync } = require('fs');
```
Comment (prudhvicbh) [inline .github/actions/auto-version-bump/index.js:3]: Our preferred language is typescript, Is that easy to port this into typescript? May just change  .js to .ts

File: .github/actions/auto-version-bump/index.js (hunk 2 above)
Comment (ddhorstman) [reply .github/actions/auto-version-bump/index.js:2]: I tried to include the tests, but I couldn't get them to work. They seem to require some extra configuration (injecting github credentials & a repo link) that wasn't documented.

File: .github/actions/auto-version-bump/index.js (hunk 3 above)
Comment (ddhorstman) [reply .github/actions/auto-version-bump/index.js:3]: I prefer typescript too, but in this case it's not just a filename change - you also have to add a bunch more configuration and customize `action.yml` in ways I'm unfamiliar with. I'd much rather plan to come back and refactor this to typescript at a later date if we find ourselves updating it a lot.

File: .github/actions/auto-version-bump/index.js (hunk 1 above)
Comment (ddhorstman) [reply .github/actions/auto-version-bump/index.js:1]: The base source code appears to be pretty well covered by the author's original tests (I assume - I couldn't get them to work without the extra undocumented config vars I mentioned below).

I tested my changes by [running the workflow several times](https://github.com/ClipboardHealth/test-chassis/actions). I originally planned to add some unit tests for my additions too, but after the included tests gave me trouble, I fell back on manual testing.

**Review by pgbezerra — COMMENTED** [review]
My understanding of this code is that we need to bump the version when a DependaBot merges a PR a package upgrade.

If we use `semantic-release` we will have a more powerful versioning flow, as you can see in this repo and in the [terraform-clipboardstaffing-modules](https://github.com/ClipboardHealth/terraform-clipboardstaffing-modules).

> The documentation is better in the terraform modules repository. Please check the contribution guide there.

To accomplish that, you can:

- set a type prefix in DependaBot `commit-message` as [the documentation](https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#commit-message) says.
- Configure `.commitlint.config.json` with your customized type
- Configure `.releaserc.json` to bump a version with that customized type
- Enable the semantic-release pipeline

## Why semantic-release

- It is a fully automated version management and package publishing
- Generating the release notes automatically
- Allows customization
- Fits perfectly with GitHub:
  - Create the release
  - Create labels, and comment in the PRs released
    image
- Don't need customized coding to do the basics, just configuration
- We can create plugins to customize behavior, but I don't believe we will need it
- Publish NPM if needed

**Comment by ddhorstman** [comment]
> To accomplish that, you can set a type prefix in DependaBot commit-message as [the documentation](https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#commit-message) says.

@pgbezerra Unfortunately, this won't cover our use-case, as that prefix applies the same to all dependabot PRs. As far as I can tell, you can't specify different prefixes for different version changes (`fix` `feat` `feat!`), which is what we'd need to synchronize version bumps between the main package and its dependencies.

Instead, it would be necessary to [write a plugin](https://semantic-release.gitbook.io/semantic-release/developer-guide/plugin) to support our use-case, which would require a good amount of re-work on a task that has already been completed to spec. @nonameable Do you think the benefits listed justify abandoning the GitHub Actions approach and moving to `semantic-release` by writing a plugin?

**Comment by pgbezerra** [comment]
@nonameable, you talked to me about this in our sync and I didn't realize it at that time, but I don't think we should continue with this approach.

According to `semver` major versions are bumped only when a breaking change is implemented.
If a service didn't change its architecture deeply, or if the contract didn't change, no major version should be bumped.

But before insisting on this, let me ask:

Which issue is solved by bumping the major version when a dependency is upgraded?

@mikeacook, you have created a [ticket](https://clipboardhealth.atlassian.net/browse/TK-279) in the Kraken board talking about bump versioning and `semver`.

The discussion is still open and I think we are anticipating the implementation without having it well-defined.

> I'll link this PR on the card.

**Comment by pgbezerra** [comment]
@mikeacook,

@mitre-tm and I started writing a document to define a rule for versioning company-wide.

It's a WIP document and it was not published yet. Discussions are welcome.
https://clipboardhealth.atlassian.net/wiki/spaces/EPP/pages/edit-v2/2706538693?draftShareId=afbb18c8-8e00-446f-9a67-fbca848e726d&inEditorTemplatesPanel=auto_closed

**Comment by ddhorstman** [comment]
> Which issue is solved by bumping the major version when a dependency is upgraded?

If a dependency included a major version bump, that means there was a backwards incompatible-change. To avoid downstream users of the chassis accidentally introducing breaking changes, we want to increment the major version of the chassis in sync.

**Comment by mitre-tm** [comment]
> > Which issue is solved by bumping the major version when a dependency is upgraded?
> 
> If a dependency included a major version bump, that means there was a backwards incompatible-change. To avoid downstream users of the chassis accidentally introducing breaking changes, we want to increment the major version of the chassis in sync.

Is the plan for services to have implicit direct dependencies on their chassis-transitive dependencies?

Can someone point to the Tech Design Doc and proof-of-concept of the chassis?

**Comment by ddhorstman** [comment]
> Is the plan for services to have implicit direct dependencies on their chassis-transitive dependencies?
Yes. This is what motivates us to synchronize major version bumps between the "chassis-transitive" dependencies and the chassis itself.

> Could someone please share here the Tech Design Doc and proof-of-concept of the chassis? cc @nonameable @mikeacook
This task is currently in the exploratory phase. We are working on a proof-of-concept - the custom action in this PR is part of that planned proof-of-concept.

**Comment by mitre-tm** [comment]
> > Is the plan for services to have implicit direct dependencies on their chassis-transitive dependencies?
> 
> Yes. This is why we plan to to synchronize major version bumps between the "chassis-transitive" dependencies and the chassis itself.

I generally tend to think implicit direct transitive dependencies are a bad idea, especially if the depended-upon packages are transitive from a package outside the service repo (i.e., not from a [workspace](https://docs.npmjs.com/cli/v8/using-npm/workspaces)). By the way, it will cause [depcheck](https://www.npmjs.com/package/depcheck) to fail. Also, a breaking change in a chassis dependency that is not directly used by the service is _not_ a breaking change for the service.


> > Could someone please share here the Tech Design Doc and proof-of-concept of the chassis?
> 
> This task is currently in the exploratory phase. We are working on a proof-of-concept - the custom action in this PR is part of that planned proof-of-concept.

If the task is currently in the exploratory phase, we should definitely not be working on automating things that we are not even sure we will need. cc @nonameable

**Comment by nonameable** [comment]
First of all, thanks for both your input @mitre-tm and for sharing the links above, that way we can all have a clear understanding of the terms, nevertheless,  I don't think we are there yet. You asked:

> Is the plan for services to have implicit direct dependencies on their chassis-transitive dependencies?

I know @ddhorstman said yes. Before he answered we tried to figure out with the devX team you meant by  _implicit direct dependencies on their chassis-transitive dependencies_ for a moment and that's why the answer was yes, but after reading the next comment I think we didn't have a right understanding of the concept. Now that I understand it better:

> I generally tend to think implicit direct transitive dependencies are a bad idea...

Agreed, and I dare to say that everyone in the devX team would say the same. 

> Follow the rules to be safe:
> 
> Do not use transitive deps implicitly: add a direct dependency for transitive deps you re-use in your code.
> [Dependency Management — Dataverse.org](https://guides.dataverse.org/en/latest/developers/dependencies.html)
> 
> By the way, implicit depdencies will cause [depcheck](https://www.npmjs.com/package/depcheck) to fail due to "Missing dependencies".

Agreed again. 

With that in mind, I want to change the answer that @ddhorstman gave to a no, since the chassis is not going to behave like the dependency graph in the link you shared. The behavior we are aiming for is for the chassis to re-export the functionality of the 2nd order dependencies. Although the 2nd order dependencies clearly will be dependencies of the chassis, services would not have direct access to those anymore. Instead, they would access to that same functionality via the chassis itself, and for them the 2nd order dependency functionality would be treated more as a module that the chassis offers more than yet another dependency. Does that make sense? @ClipboardHealth/devx-team ? Please correct me if I'm wrong. For a clearer picture of what we are aiming for, read the _How are we solving the problem?_ section of [the memo](https://docs.google.com/document/d/18Z0DPepz3Vq7MDtoVwjLjiN7vW2vC78NFmyHiaaJ_Zc/edit#). 

With this picture in mind, what is your opinion @mitre-tm? 

>Finally, a breaking change in a chassis dependency that is not directly used by the service is not something of which the service needs to be aware, and false positives on breaking changes decrease the [signal-to-noise ratio](https://en.wikipedia.org/wiki/Signal-to-noise_ratio).

I agree with this @mitre-tm . That is a con of the approach we are proposing here, am I right @ClipboardHealth/devx-team ?


> >Would someone please share here the Tech Design Doc and proof-of-concept of the chassis?
> 
> >This task is currently in the exploratory phase. We are working on a proof-of-concept - the custom action in this PR is part of that planned proof-of-concept.
> 
> If the task is currently in the exploratory phase, we should definitely not be working on automating things that we are not even sure we will need. cc @nonameable
Given the behavior we want the chassis to have, we realized that the semantic versioning approach would not completely fit our use case, since unlike regular dependency management, given that the chassis is "re-exporting" the functionality, a breaking change in a chassis dependency that is directly used by the service is something of which the service needs to be aware (opposite case of what you mentioned before). That is the reason of being of this PR which tries to address that, since we believe we do need it. If you or @pgbezerra see a way in which we can extend the semantic versioning approach to support this, we are more than happy to try it! We want to explore different ways of doing this. @ddhorstman suggested a plug-in, but let me know your thoughts. 

Finally, maybe the way that we are proposing the chassis to behave is not the best, and that's why we proposed a way in the memo, so that people could give us early feedback on that and fresh perspectives. Is never too late to do the best thing though. Please share other options that you'd consider are better.

**Comment by sergelerator** [comment]
There is a lot to unpack here, hopefully I am understanding things correctly when I'm answering the outstanding issues.

---

@mitre-tm wrote:
> I generally tend to think implicit direct transitive dependencies are a bad idea, especially if the depended-upon packages are transitive from a package outside the service repo

I agree with this. However, as @nonameable pointed out, application code that depends on the chassis will not directly import dependencies from transitive dependencies, or at least that is not the intended way to use it. You can take a look at the proposal on how to import chassis components [here](https://github.com/ClipboardHealth/nestjs-chassis/tree/feat/DEVX-234-bootstrapping#usage), copying a sample, the intended way to import chassis components is:

```ts
import { AuthGuard } from "@clipboard-health/nestjs-chassis";
// or
import { AuthGuard } from "@clipboard-health/nestjs-chassis/lib/auth-guard";
```

With `@clipboard-health/nestjs-chassis` being a direct explicit dependency.

---

@nonameable wrote:
> Finally, maybe the way that we are proposing the chassis to behave is not the best, and that's why we proposed a way in the memo, so that people could give us early feedback on that and fresh perspectives. Is never too late to do the best thing though. Please share other options that you'd consider are better.

I echo @nonameable's sentiment here, the chassis serving as an aggregator for multiple tools that have already been developed (and possibly other tools that will continue to be developed) is the solution we came up with for reducing the amount of manual work to keep CBH's internal dependencies up to date across the growing list of microservices and proposed it in [the memo](https://docs.google.com/document/d/18Z0DPepz3Vq7MDtoVwjLjiN7vW2vC78NFmyHiaaJ_Zc/), if this is not a good solution (and we're all ears for arguments that challenge our solution's validity!), then let's re-focus our efforts.

**Comment by mitre-tm** [comment]
> I echo @nonameable's sentiment here, the chassis serving as an aggregator for multiple tools that have already been developed (and possibly other tools that will continue to be developed) is the solution we came up with for reducing the amount of manual work to keep CBH's internal dependencies up to date across the growing list of microservices and proposed it in [the memo](https://docs.google.com/document/d/18Z0DPepz3Vq7MDtoVwjLjiN7vW2vC78NFmyHiaaJ_Zc/), if this is not a good solution (and we're all ears for arguments that challenge our solution's validity!), then let's re-focus our efforts.

Keeping dependencies up-to-date, be them internal or external packages, is a [general problem](https://clipboardhealth.slack.com/archives/C03LY57G5MF/p1666250706818489), of which just address a small part (internal libraries) the chassis would address, leading to a hybrid approach which would increase cognitive load. Also, by bundling multiple updates at once, we would be accumulating risk (big bang vs soft-rollout, changing one thing at a time vs changing a bunch of things all at once).

The creation of Terraform modules and GitHub custom Actions / reusable workflows will effectively make them libraries, so we reduce the problem to one of semi-automating the update of dependencies, which is the general problem I alluded above.

Earlier today, I have discussed with @mikeacook about investing in a more general solution to the problem, be it using Dependabot, Renovate or some other tool, instead of the microservice chassis.

### References:
- [Dependabot ❤️s private dependencies | The GitHub Blog](https://github.blog/2021-03-15-dependabot-private-dependencies/)
- [Configuring Dependabot version updates - GitHub Docs](https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuring-dependabot-version-updates)
- [Renovate · GitHub Marketplace](https://github.com/marketplace/renovate)
- [build: replace npm install by npm clean-install for increased reproducibility and security](https://github.com/ClipboardHealth/microservice-templates/pull/63)
- relevant Slack thread: https://clipboardhealth.slack.com/archives/C03GV3J181X/p1660925223944709
- [We’re drowning | snarfed.org](https://snarfed.org/2022-03-10_were-drowning-software-dependencies) - also [shared](https://clipboardhealth.slack.com/archives/C03LY57G5MF/p1666250706818489) in #tech-watercooler, with a link for HN discussion

**Pull request closed without merging**